### PR TITLE
[#158818736] Add callback to intercept SAML requests and responses

### PR DESCRIPTION
**Pull request by AleDore**
Added callback log for Spid audit purposes.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -72,7 +73,13 @@ export { noopCacheProvider, IServiceProviderConfig, SamlConfig };
 const withSpidAuthMiddleware = (
   acs: AssertionConsumerServiceT,
   clientLoginRedirectionUrl: string,
-  clientErrorRedirectionUrl: string
+  clientErrorRedirectionUrl: string,
+  logCallback?: (
+    sourceIp: string | null,
+    payload: string,
+    timestamp: string,
+    isRequest: boolean
```
Comment (gunzip) [inline src/index.ts]: this should be a literal with type
`type CALLBACK_PAYLOAD_TYPE = "REQUEST" | "RESPONSE"`

avoid using booleans as parameters, they do not scale and makes calls obscure

File: src/index.ts [hunk 2]
```diff
@@ -107,6 +114,14 @@ const withSpidAuthMiddleware = (
         );
         return res.redirect(clientLoginRedirectionUrl);
       }
+      if (logCallback && isSome(maybeDoc)) {
+        logCallback(
+          requestIp.getClientIp(req),
+          maybeDoc.toString(),
```
Comment (gunzip) [inline src/index.ts]: i think this should be `maybeDoc.value` but `req.body` is enough

File: src/index.ts [hunk 3]
```diff
@@ -117,14 +132,21 @@ const withSpidAuthMiddleware = (
...
 export function withSpid(
   appConfig: IApplicationConfig,
   samlConfig: SamlConfig,
   serviceProviderConfig: IServiceProviderConfig,
   redisClient: RedisClient,
   app: express.Express,
   acs: AssertionConsumerServiceT,
-  logout: LogoutT
+  logout: LogoutT,
+  logCallback?: (
```
Comment (BurnedMarshal) [inline src/index.ts]: You need pass this function even to `makeSpidStrategy`

File: src/index.ts [hunk 4]
```diff
@@ -72,7 +73,13 @@ export { noopCacheProvider, IServiceProviderConfig, SamlConfig };
 const withSpidAuthMiddleware = (
   acs: AssertionConsumerServiceT,
   clientLoginRedirectionUrl: string,
-  clientErrorRedirectionUrl: string
+  clientErrorRedirectionUrl: string,
+  logCallback?: (
```
Comment (gunzip) [inline src/index.ts]: naming: you don't know how this callback will be used by the caller

```suggestion
  callback?: (
```

File: src/index.ts [hunk 5]
```diff
@@ -117,14 +132,21 @@ const withSpidAuthMiddleware = (
...
   app: express.Express,
   acs: AssertionConsumerServiceT,
-  logout: LogoutT
+  logout: LogoutT,
+  logCallback?: (
+    sourceIp: string | null,
+    payload: string,
+    timestamp: string,
+    isRequest: boolean
+  ) => void
```
Comment (gunzip) [inline src/index.ts]: can you make a type for this method ie. `WithSpidCallbackT` and export it ?
(so the caller can use the exported type if she wants to define the callback outside withspid())

File: src/index.ts [hunk 6]
```diff
@@ -192,9 +214,13 @@ export function withSpid(
...
 
-      const spidAuth = passport.authenticate("spid", {
-        session: false
-      });
+      const spidAuth = passport.authenticate(
+        "spid",
+        {
+          session: false
+        },
+        logCallback
```
Comment (gunzip) [inline src/index.ts]: why is this needed ?

File: src/strategy/saml_client.ts [hunk 7]
```diff
@@ -10,7 +11,13 @@ export class CustomSamlClient extends PassportSaml.SAML {
...
     private extededCacheProvider: IExtendedCacheProvider,
     private tamperAuthorizeRequest?: XmlTamperer,
-    private preValidateResponse?: PreValidateResponseT
+    private preValidateResponse?: PreValidateResponseT,
+    private logCallback?: (
+      sourceIp: string | null,
+      payload: string,
+      timestamp: string,
+      isRequest: boolean
+    ) => void
```
Comment (gunzip) [inline src/strategy/saml_client.ts]: use the type definition, see previous comment

File: src/strategy/saml_client.ts [hunk 8]
```diff
@@ -70,9 +77,17 @@ export class CustomSamlClient extends PassportSaml.SAML {
...
-              )
+              .chain(tamperedXml => {
+                if (this.logCallback !== undefined) {
+                  this.logCallback(
+                    requestIp.getClientIp(req),
+                    tamperedXml,
+                    new Date().toISOString(),
+                    true
+                  );
+                }
```
Comment (gunzip) [inline src/strategy/saml_client.ts]: this is not a good place to put the callback since this code runs only if tamperAuthorizeRequest is defined

File: src/strategy/spid.ts [hunk 9]
```diff
@@ -44,7 +44,13 @@ export class SpidStrategy extends SamlStrategy {
...
     private tamperAuthorizeRequest?: XmlTamperer,
     private tamperMetadata?: XmlTamperer,
-    private preValidateResponse?: PreValidateResponseT
+    private preValidateResponse?: PreValidateResponseT,
+    private logCallback?: (
+      sourceIp: string | null,
+      payload: string,
+      timestamp: string,
+      isRequest: boolean
+    ) => void
```
Comment (gunzip) [inline src/strategy/spid.ts]: use the type definition

File: src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (AleDore) [reply src/index.ts]: sorry my bad. It is a misprint :)

File: src/index.ts (hunk 3 above)
Comment (AleDore) [reply src/index.ts]: Yes, I'll make changes!

File: src/strategy/saml_client.ts (hunk 8 above)
Comment (AleDore) [reply src/strategy/saml_client.ts]: Do you mean I should use `xml` before it is tampered?

File: src/index.ts [hunk 10]
```diff
@@ -51,6 +52,14 @@ export type AssertionConsumerServiceT = (
 // logout express handler
 export type LogoutT = () => Promise<IResponsePermanentRedirect>;
 
+export type CallbackPayloadType = "REQUEST" | "RESPONSE";
+
+export type WithSpidCallbackT = (
+  sourceIp: string | null,
+  payload: string,
+  timestamp: string,
```
Comment (gunzip) [inline src/index.ts]: it looke like this is more a date than a timestamp
```suggestion
  createdAt: UTCISODateFromString,
```

File: src/index.ts [hunk 11]
```diff
@@ -107,6 +118,14 @@ const withSpidAuthMiddleware = (
         );
         return res.redirect(clientLoginRedirectionUrl);
       }
+      if (callback && isSome(maybeDoc)) {
+        callback(
+          requestIp.getClientIp(req),
+          req.body,
+          UTCISODateFromString.decode(new Date()).getOrElse(new Date()),
```
Comment (gunzip) [inline src/index.ts]:
```suggestion
         new Date(),
```

File: src/strategy/saml_client.ts [hunk 12]
```diff
@@ -70,9 +74,19 @@ export class CustomSamlClient extends PassportSaml.SAML {
...
-                this.extededCacheProvider.save(tamperedXml, this.config)
-              )
+              .chain(tamperedXml => {
+                if (this.callback !== undefined) {
+                  this.callback(
+                    requestIp.getClientIp(req),
+                    tamperedXml,
+                    UTCISODateFromString.decode(new Date()).getOrElse(
+                      new Date()
+                    ),
```
Comment (gunzip) [inline src/strategy/saml_client.ts]:
```suggestion
                      new Date(),
```

File: src/index.ts [hunk 13]
```diff
@@ -51,6 +53,14 @@ export type AssertionConsumerServiceT = (
 // logout express handler
 export type LogoutT = () => Promise<IResponsePermanentRedirect>;
 
+export type CallbackPayloadType = "REQUEST" | "RESPONSE";
+
+export type WithSpidCallbackT = (
+  sourceIp: string | null,
```
Comment (gunzip) [inline src/index.ts]: why null ?
